### ISSUE TITLE
Adding some explanations to example for CustomLayer with Threejs

### ISSUE THREAD

**Issue by MichaelLangbein** [maplibre/maplibre-gl-js#3427]
## User Story

As a person relatively new to maplibre, I'd like to have some additional information about how maplibre's terrain interacts with `CustomLayer`s projection onto the map, so that I can, for example, place threejs-objects correctly on 3d-terrain.

## Rationale

This is really just a request for some more documentation of existing features. 

There is already some nice documentation on using `CustomLayers` with threejs [here](https://maplibre.org/maplibre-gl-js/docs/examples/add-3d-model/). 
But this example doesn't make use of maplibre's 3d-terrain. If one enables the terrain, threejs-objects can sink into or hover above the ground. 
This issue has been documented by [daniel-f](https://stackoverflow.com/users/277267/daniel-f) in [this stackoverflow question](https://stackoverflow.com/questions/73409827/maplibre-gl-js-with-terrain-layer-how-to-pin-a-horizontal-plane-to-specific-alt).

I think this gif from daniel-f's post describes the issue quite clearly. What's being displayed is a threejs-plane which is located at sea-level, but seems to move up to always match the center's elevation.
![0C4Ge](https://github.com/maplibre/maplibre-gl-js/assets/13807050/2008dc70-2b65-4f92-9f66-a1f8f0fb9375)

I've provided one possible approach to solving the issue in [this simple stackblitz](https://stackblitz.com/edit/typescript-mtbmkh?file=index.ts). 

Essentially, it seems to me that the confusion is caused by the following:
- `CustomLayer.render(gl, matrix)` is being passed the `Transform.mercatorMatrix`, but this matrix doesn't yet account for terrain elevation
- Elevation is taken into account later: I think (please correct me if I'm wrong here!) that maplibre uses `Transform.projMatrix` ([here](https://github.com/maplibre/maplibre-gl-js/blob/8edbac09725f807aae91009d31c30c3d5168b066/src/geo/transform.ts#L859C1-L869C1)) when actually displaying the `CustomLayer`.
- This has the effect that all Objects of the given threejs-scene are displayed relative to `elevation`, which is the height of `center` - even if they are actually on a point of the terrain that lies above or below `center`.
- So if I want to account for the difference in elevation between `center` and my model, I could do something like this (snippet copied from the stackblitz, `modelTransform` analogous to the [already existing example](https://maplibre.org/maplibre-gl-js/docs/examples/add-3d-model/)):

```ts
const modelOrigin = new LngLat(11.53, 47.67);
const modelElevation = map.terrain ? map.terrain.getElevationForLngLatZoom(modelOrigin, map.getZoom()) : 0;
const modelOriginMercator = mlg.MercatorCoordinate.fromLngLat(modelOrigin, 0);

const terrainCenterElevation = map.transform.elevation;
const deltaMetersZ = modelElevation - terrainCenterElevation;
const mercatorPerMeter =modelOriginMercator.meterInMercatorCoordinateUnits();
const deltaMercatorZ = deltaMetersZ * mercatorPerMeter;

const modelTransform = {
  translateX: modelOriginMercator.x,
  translateY: modelOriginMercator.y,
  translateZ: modelOriginMercator.z + deltaMercatorZ,
  rotateX: Math.PI / 2,
  rotateY: 0,
  rotateZ: 0,
  scale: modelOriginMercator.meterInMercatorCoordinateUnits(),
};
```

It might at this point also make sense to document the relevant variables. I'm very new to maplibre, but my understanding is:

- centerPoint [pixel]: the viewport's center
- center [Long, Lat]: the viewport's center projected onto a point on terrain
- elevation [meter]: the height of center relative to sea-level

## Impact

Adding some information on how maplibre accounts for terrain and on how to make use of that inside of a CustomLayer would probably make getting started a bit easier for people coming from other 3d-libraries. 

Of course, this holds only if there is not a more elegant way to account for objects being situated on a different terrain-level than `center`! If there is, I'd love to hear about it!

**Comment by HarelM** [comment]
Any improvements to the docs are welcomed!
I'm not very good with ThreeJS and how it works, especially with all the terrain matrices etc.
If you think you can add an example of ThreeJS with terrain showing things in different height it will be great!
The examples are very easy to create - it's a simple html file with the relevant code.
If you think a guide will be more appropriate here, you can add a guide to the docs site.
I'll be happy to review any PR you'll open.
If you need information on how to setup stuff let me know, although this repo is filled with relevant readme files that should do the trick.

Relevant links (maybe):
- https://github.com/maplibre/maplibre-gl-js/issues/2684
- https://github.com/maplibre/maplibre-gl-js/discussions/1523
- https://github.com/maplibre/maplibre-gl-js/discussions/1548

**Comment by neodescis** [comment]
I went through this recently myself, and an example would be great. FYI, this is also related to (closed) issue #1654. The issue has a link to a ["debug" page](https://github.com/maplibre/maplibre-gl-js/blob/7bab10e8dfd1680471126b02b4849423a5f50365/test/debug-pages/terrain-3d-object.html). I'm not sure why that page didn't make its way to an official example, but it's worth looking at.

**Comment by neodescis** [comment]
Makes sense. I posted it mostly because it got me up and running, and I thought it might be good for @MichaelLangbein to look at it.

**Comment by HarelM** [comment]
Resolved by #3429
